document how to build under alpine

deitch opened the pull request:
I had to build under alpine (musl), so I documented it. Hopefully this is helpful for others.

lifubang commented on docs/alpine.md:
````diff
@@ -0,0 +1,19 @@
+To build under alpine linux, whether booted alpine or in a container:
+
+```sh
+apk add make gcc libseccomp-dev musl-dev linux-headers git
````
1. Is ‘git’ really necessary?
2. Even in the main Readme, there is no introduction to tell users to install gcc or musl, I think it’s the basic info that we should know. So if we want add some information about pre-requirements of build, the best place should be ‘../README.md’.
3. This file is not referenced by any other files.
WDYT?

deitch replied:
Yeah git is necessary. Go uses it under the covers as far as I know. I ran `make runc` and got an error for missing git. 

Sure. Maybe this whole doc should be "prerequisites.md", cover both glibc and musl based OSes, and linked from README?

I'm more than happy to change this around as needed, as long as I know it will get accepted at the end.

Thoughts?

lifubang replied:
For my personal, I think we should add ‘pre-requirements.md’, because if we clone the repository to local for the first time, we usually can’t compile runc successfully in a big possibility, especially in GitHub online codespace. But we should consider other maintainers’ option.

cyphar replied:
`git` isn't necessary for building. Our `Makefile` does use `git` to get the commit information to embed in the binary, but you can do `make COMMIT=""' to disable that. Go doesn't require `git` either (we vendor our dependencies, so you can build offline).

deitch replied:
Ah, ok. I didn't pay close enough attention to realize the deps had been vendored, so git would not be necessary.

Although, if the normal `make` path expects git, then I think listing git as a dependency makes sense. I can always add a comment about leaving it out if you do not include the commit as `make COMMIT=""`.

In any case, what did we decide here? Do we want to make this a generic "prerequirements.md" or "build-requirements.md" or even just "building.md"/"build.md" and link from the main README.md?

rata replied:
@lifubang @cyphar wanna take another look at the new version?

AkihiroSuda reviewed (changes requested):
This should be merged to https://github.com/opencontainers/runc/blob/main/README.md#building

deitch commented:
OK, I merged all of this into the section in README that @AkihiroSuda recommended, added apt and yum commands for centOS, and moved the comment about optional seccomp to that section.

Ready for next run of review.

rata commented on README.md:
````diff
@@ -29,12 +29,41 @@ A third party security audit was performed by Cure53, you can see the full repor
+On Alpine Linux, you can install the required dependencies with:
+
+```bash
+apk --update add make gcc libseccomp-dev musl-dev linux-headers git
````
Why is alpine the only one with linux.headers dep?

deitch replied:
I don't actually know. Either I missed something, or it needs to add deps on Alpine which are there already on the others. That latter doesn't sound right to me, but I will run a clean test.

deitch replied:
Ran clean tests.

* `alpine:3.20`: I missed `bash`, added that to the doc. It builds cleanly now, both `make runc` and `make static`
* `ubuntu:22.04`:  I didn't get as far as the headers question, because gcc errors out with `gcc: error: unrecognized command-line option '-m64'`. Running gcc 11.4.0.
* `fedora:40`: same as ubuntu. I didn't get as far as the headers question, because gcc errors out with `gcc: error: unrecognized command-line option '-m64'`. Running gcc 14.1.1

In all cases, running go 1.22.5.

I _did_ try inside the golang container from the `Dockerfile` in this repo: `golang:1.22-bookworm`. I do not get the same gcc errors. In there, it is gcc 12.2.0.

What obvious thing am I missing?

rata replied:
Are you cross-compiling? It seems like https://github.com/opencontainers/runc/blob/main/cc_platform.mk is causing issues again.

Does it work if you comment its include in the Makefile?

deitch replied:
> Are you cross-compiling? 

Not intentionally. 😄 

> Does it work if you comment its include in the Makefile?

Same error:

```sh
root@df3feb358386:/pwd# make static
rm -f libcontainer/dmz/binary/runc-dmz
go generate -tags "seccomp urfave_cli_no_docs " ./libcontainer/dmz
make[1]: Entering directory '/pwd/libcontainer/dmz'
gcc -fno-asynchronous-unwind-tables -fno-ident -s -Os -nostdlib -lgcc -static -o binary/runc-dmz _dmz.c
strip -gs binary/runc-dmz
make[1]: Leaving directory '/pwd/libcontainer/dmz'
go build -trimpath -buildmode=pie  -tags "seccomp urfave_cli_no_docs  netgo osusergo" -ldflags "-X main.gitCommit=v1.2.0-rc.2-22-g419ef873 -X main.version=1.2.0-rc.2+dev -linkmode external -extldflags -static-pie " -o runc .
# runtime/cgo
gcc: error: unrecognized command-line option '-m64'
make: *** [Makefile:105: static-bin] Error 1
```

Or if I do it explicitly:

```sh
root@df3feb358386:/pwd# make -n static
rm -f libcontainer/dmz/binary/runc-dmz
go generate -tags "seccomp urfave_cli_no_docs " ./libcontainer/dmz
go build -trimpath -buildmode=pie  -tags "seccomp urfave_cli_no_docs  netgo osusergo" -ldflags "-X main.gitCommit=v1.2.0-rc.2-22-g419ef873 -X main.version=1.2.0-rc.2+dev -linkmode external -extldflags -static-pie " -o runc .
if test -s libcontainer/dmz/binary/runc-dmz; then \
        set -Eeuo pipefail; \
        export LC_ALL=C; \
        diff -u \
                <(readelf -h runc | grep -E "(Machine|Flags):") \
                <(readelf -h libcontainer/dmz/binary/runc-dmz | grep -E "(Machine|Flags):"); \
fi


root@df3feb358386:/pwd# go build -trimpath -buildmode=pie  -tags "seccomp urfave_cli_no_docs  netgo osusergo" -ldflags "-X main.gitCommit=v1.2.0-rc.2-22-g419ef873 -X main.version=1.2.0-rc.2+dev -linkmode external -extldflags -static-pie " -o runc .
# runtime/cgo
gcc: error: unrecognized command-line option '-m64'
```

deitch replied:
The `golang:1.22-bookworm`, which does not generate the error, has the same command-line. 🤷‍♂️

rata commented on README.md:
````diff
@@ -29,12 +29,41 @@ A third party security audit was performed by Cure53, you can see the full repor
+
+In addition to Go, building `runc` requires multiple utilities and libraries to be installed on your system.
+
+On Ubuntu/Debian, you can install the required dependencies with:
+
+```bash
+apt update && apt install -y make gcc libseccomp-dev pkg-config git
+```
+
+On CentOS/Fedora, you can install the required dependencies with:
+
+```bash
+yum install -y make gcc libseccomp-devel pkg-config git
````
Can you verify building with only these on the respective container (like one for debian, another for ubuntu, fedora, etc.)?

It feels untested, given that alpine has linux-headers and here we don't. Am I missing something?

deitch replied:
Yup, as above, I will run a clean test.

rata replied:
Yes, all works fine today in CI, locally, etc.

deitch replied:
So why isn't it working in a fresh ubuntu or fedora container? Is it a go configuration thing? A gcc configuration thing? Something else?

Can you try it?

```sh
docker run -it --rm -v $(pwd):/src -w /src ubuntu:22.04
```

Install go 1.22.5, install the various packages, in this case:

```sh
apt update && apt install -y make gcc libseccomp-dev pkg-config git
```

and then `make runc` or `make static`.

rata replied:
Works fine here. I installed bash and golang from debian, and then installed go 1.22.5 from here (following this: https://go.dev/doc/install) because debian's version was too old.

deitch replied:
You did it in a debian container? Which tag? Do you mind pasting the output?

deitch replied:
I just recreated the same issue on `debian:sid-slim`. It is something common that you have set and I do not, or vice-versa.

rata replied:
No, I did it in ubuntu, using the line you told me to use.

Here is the info about the ubuntu image:
```
$ docker inspect  08d22c0ceb15
[
    {
        "Id": "sha256:08d22c0ceb150ddeb2237c5fa3129c0183f3cc6f5eeb2e7aa4016da3ad02140a",
        "RepoTags": [
            "ubuntu:22.04"
        ],
        "RepoDigests": [
            "ubuntu@sha256:67211c14fa74f070d27cc59d69a7fa9aeff8e28ea118ef3babc295a0428a6d21"
        ],
```

Let me know if you need anything else to debug.

deitch replied:
This is really strange. I even went so far as to replace `gcc` with a script that dumps the call to `/tmp/gcc.out` just so I can see what is called differently.

deitch replied:
OK, looks like you were onto something, but not sure. I ran the command with `go build -x`. What surprised me was this output:

```sh
go build -x -trimpath "-buildmode=pie"  -tags "seccomp urfave_cli_no_docs " -ldflags "-X main.gitCommit=v1.2.0-rc.2-22-g41
9ef873 -X main.version=1.2.0-rc.2+dev " -o runc .
WORK=/tmp/go-build2257789363
cd /pwd
git status --porcelain
cd /pwd
git -c log.showsignature=false show -s --format=%H:%ct
mkdir -p $WORK/b141/
cd /usr/local/go/src/runtime/cgo
TERM='dumb' CGO_LDFLAGS='"-O2" "-g" "-lpthread"' /usr/local/go/pkg/tool/linux_amd64/cgo -objdir $WORK/b141/ -importpath runtime/cgo -import_runtime_cgo=false -import_syscall=false -- -I $WORK/b141/ -O2 -g -Wall -Werror -fno-stack-protector ./cgo.go
# runtime/cgo
gcc-13: error: unrecognized command-line option '-m64'
```

Why would it be running `/usr/local/go/pkg/tool/linux_amd64/cgo`? That is for building for amd64. Yet, I am running on aarch64. 

GOARCH is unset. If I execute it with `GOARCH=arm64`, then it works, e.g.:

```
make static GOARCH=arm64
```

Oh crap! It is my fault! I copied the wrong link for installing go, had the two of them - amd64 and arm64 - and copied the wrong one! 🤯

deitch replied:
I can confirm that the builds are not an issue now. Alpine needs `linux-headers` added explicitly, the others do not.

deitch replied:
Yeah, and what a silly mystery. Sigh. Oh well, may all our problems be that easy to solve.

> We are not targetting building inside a container (or not only), I think all the dependencies should be listed. IMHO we mention the respective package for each distro.

Agreed. It was just easier to test the various distros in the container.

No, the system doesn't come with it:

```sh
$ docker run -it --rm -v $(pwd):/pwd -w /pwd ubuntu:22.04
# ls -l /usr/include/linux/stat.h
ls: cannot access '/usr/include/linux/stat.h': No such file or directory
```

However, installing gcc causes it to exist:

```sh
# apt install -y gcc
...
# ls -l /usr/include/linux/stat.h 
-rw-r--r-- 1 root root 6726 Jul  1 09:13 /usr/include/linux/stat.h
# dpkg-query -S /usr/include/linux/stat.h
linux-libc-dev:arm64: /usr/include/linux/stat.h
```

That isn't going to be fun to track down, but I don't think we need to. As long as we explicitly say you need `linux-libc-dev` - which makes sense, and parallels `musl-dev` on alpine - we should be good.

For fedora, it also is `gcc`, which installs `kernel-headers`, so we can do that package there.

```sh
# rpm -qf /usr/include/linux/stat.h
kernel-headers-6.9.4-200.fc40.aarch64
```

I updated the doc to reflect it. Take a look.

rata replied:
Thanks!

rata commented on README.md:
````diff
@@ -118,7 +146,7 @@ You can run a test using your container engine's flags by setting `CONTAINER_ENG
 ```
 
-### Dependencies Management
+### Go Dependencies Management
````
I'd lean towards not changing this. If at some point we have non-go dependencies (we do have a small C dep, nolibc), we want to mention them in the README and we want a different section, we can do this change. But I'm unsure what it adds now.

deitch replied:
Because this section talks only about go dependencies. All of the other are handled by package management further up (or go itself, which is separate). To have this section be called "dependencies management", but then have all of the dependencies already listed and this section only discusses go, can be confusing.

rata replied:
Ohhh, okay, I completely missed that! I see the point now, sorry!

rata reviewed (commented):
@deitch thanks for the PR! I've left some comments.

rata commented on README.md:
````diff
@@ -29,12 +29,41 @@ A third party security audit was performed by Cure53, you can see the full repor
+On CentOS/Fedora, you can install the required dependencies with:
+
+```bash
+yum install -y make gcc kernel-headers libseccomp-devel pkg-config git
````
Just to be extra-careful, kernel-headers is a meta-package that refers to the headers of the latest kernel in fedora and centos?

deitch replied:
As far as I know, yes. But I won't claim to be super expert in any package manager (well, I did once write an implementation for managing alpine apk packages in golang, which required figuring out how it did everything, but that was for a particular project in the past 😁 )

rata replied:
But I mean, this package exists and these instructions work fine on the fedora container you tested, right?

rata replied:
re: https://github.com/opencontainers/runc/pull/4353#issuecomment-2244980919

Awesome, thanks!

rata reviewed (approved):
@deitch thanks a lot for the PR! LGTM

I've tested this in debian:testing-slim too, and it works fine.

Can you confirm that you tested it in all the distros you documented how it should compile?

deitch commented:
Yes correct. I ran each of these commands on the specific distro.

lifubang reviewed (approved):
LGTM, thanks.
